src/main/java/com/ldtteam/structurize/placement/handlers/placement/PlacementHandlers.java

```diff
@@ -63,6 +97,44 @@ private PlacementHandlers()
          */
     }
 
+    public static class FluidSubstitutionPlacementHandler implements IPlacementHandler
+    {
+        BlockState existing;
+        @Override
+        public boolean canHandle(@NotNull World world, @NotNull BlockPos pos, @NotNull BlockState blockState)
+        {
+            if (blockState.getBlock() instanceof BlockFluidSubstitution)
+            {
+                existing = world.getBlockState(pos);
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public List<ItemStack> getRequiredItems(@NotNull World world, @NotNull BlockPos pos, @NotNull BlockState blockState, @Nullable CompoundNBT tileEntityData, boolean complete)
+        {
+            return new ArrayList<>();
+        }
+
+        @Override
+        public ActionProcessingResult handle(
+          @NotNull World world,
+          @NotNull BlockPos pos,
+          @NotNull BlockState blockState,
+          @Nullable CompoundNBT tileEntityData,
+          boolean complete,
+          BlockPos centerPos)
+        {
+            if (existing.hasProperty(BlockStateProperties.WATERLOGGED))
```

Comment:
missing formatting

Reply:
now you can only waterlogged water (if should contain water check)
and you really should for whatever reason query `existing` here from world and pos
and also iirc `complete `should be taken into account here too - it should place placeholder itself when pasting instead of placing (look at other placeholders)